lecture-2023-10-31-xchat/server/src/lib.rs

```diff
@@ -0,0 +1,137 @@
+use std::collections::HashMap;
+use std::error::Error;
+use std::io::ErrorKind;
+use std::net::{SocketAddr, TcpListener, TcpStream};
+use std::sync::{Arc, Mutex, MutexGuard};
+use std::thread::{spawn, sleep};
+use std::time;
+
+use shared::Message;
+
+
+type ClientMap = HashMap::<SocketAddr, TcpStream>;
+type Clients = Arc<Mutex<ClientMap>>;
+
+
+/// `start_server` is entrypoint of server. It starts main processing loop in a separate thread
+/// while main thread keeps track on managing new client connections.
+pub fn start_server(address: &str) -> Result<(), Box<dyn Error>> {
+    let client_map: ClientMap = HashMap::new();
+    let clients: Clients = Arc::new(Mutex::new(client_map));
+
+    let thread_clients = clients.clone();
+    let handle = spawn(|| chat(thread_clients) );
+
+    listen_and_accept(address, clients.clone())?;
+
+    let _ = handle.join();
+
+    Ok(())
+}
+
+
+/// `listen_and_accept` take care of connection of new client connections.
+fn listen_and_accept(address: &str, clients: Clients) -> Result<(), Box<dyn Error>> {
+    let listener = TcpListener::bind(address)?;
+
+    for stream in listener.incoming() {
+        let stream = stream;
```

Comment:
Is this statement necessary?

Reply:
Not at al. Just yesterday, I touched it a bit during playing with `thiserror` and `anyhow`... Good catch.